orttraining/orttraining/python/training/_ortmodule_execution_manager.py

```diff
@@ -0,0 +1,268 @@
+
+from . import _utils, _ortmodule_utils, _ortmodule_output_transformation as _ortmodule_io
+
+from onnxruntime.capi.onnxruntime_inference_collection import OrtValue
+from onnxruntime.capi import _pybind_state as C
+
+import io
+import inspect
+import onnx
+import onnxruntime
+import torch
+from torch.utils.cpp_extension import ROCM_HOME
+
+ONNX_OPSET_VERSION = 12
+
+class ExecutionManager(object):
+    def __init__(self, module):
+        # Original and flattened (tranformed) output module
+        self._original_module = module._base_module
+        self._flattened_output_module = module
+
+        # Exported model
+        self._model = None
+        
+        # Model after inference optimization or gradient building.
+        self._optimized_model = None
+        self._module_graph_builder = None
+        self._graph_info = None
+
+        # TrainingAgent or InferenceAgent
+        self._execution_agent = None
+
+        # Debug flags
+        self._save_onnx = False
+        self._save_onnx_prefix = ''
+
+        self._is_training = None
+
+        # Related to training graph shape inference
+        self._current_input_shape = None
+        # default execution order is priority-based for both dynamic/static shape input for now
+        # if we observe benefit of static shape, we can expose this flag to user
+        self._use_static_shape = False
+        self._input_names_require_grad = None
+        self._module_output_schema = None
+
+        # Verbosity for logging
+        self._verbosity = _ortmodule_utils.Verbosity.WARNING
+
+        # TODO: Single device support for now
+        self._device = _utils.get_device_from_module(module)
+
+        self._module_parameters = inspect.signature(
+            self._original_module.forward).parameters.values()
+
+        # TODO: remove after PyTorch ONNX exporter supports VAR_KEYWORD parameters.
+        for input_parameter in self._module_parameters:
+            if input_parameter.kind == inspect.Parameter.VAR_KEYWORD:
+                raise NotImplementedError(
+                    "The model's forward method has **kwargs parameter which is currently not supported.")
+
+        self.is_rocm_pytorch = (True if (
+            (torch.version.hip is not None) and (ROCM_HOME is not None)) else False)
+
+        # CPP extension to get torch CUDA allocator's alloc and free function addresses
+        # Disable external allocator for ROCM EP since external allocator is not supported yet.
+        self._use_external_cuda_allocator = (
+            False if self.is_rocm_pytorch else True)
+        if self._use_external_cuda_allocator:
+            self._torch_cuda_allocator = _ortmodule_utils._load_torch_allocator_cpp_extension(
+                self._verbosity)
+            self._torch_alloc = self._torch_cuda_allocator.cuda_caching_allocator_raw_alloc_address()
+            self._torch_free = self._torch_cuda_allocator.cuda_caching_allocator_raw_delete_address()
+
+    def forward(self):
```

Comment:
Maybe should we use https://docs.python.org/3.6/library/abc.html for representing abstract classes?